### PR TITLE
[Core][BlockB&S] moving application of constraints

### PR DESCRIPTION
**Pull request by philbucher**
The problem was that the application of the constraints was done in `BuildAndSolve`
This is not a problem in most cases, but it is if someone only calls `Build` (as it is done by the EigenSolver) the constraints are not applied!

=> Hence I moved the Application of the Constraints from `BuildAndSolve` to `Build`

I will add a test tmr to prevent this from happening in the future

### CODE REVIEW

**Review by RiccardoRossi — COMMENTED** [review]
why do you need dx in apply constraints??

**Comment by philbucher** [comment]
> why do you need dx in apply constraints??

I don't ...
I removed the argument, don't know why it was there to begin with. Also it is not available in `Build`

**Review by loumalouomega — DISMISSED** [review]
OK

**Review by adityaghantasala — DISMISSED** [review]

**Comment by adityaghantasala** [comment]
This is a good catch ... with this change `BuildLHS_CompleteOnFreeRows` and `BuildLHS` are also taken care of ... 

But `BuildRHS` is not ... 

By the way @RiccardoRossi   `BuildLHS_CompleteOnFreeRows` and `BuildLHS`  are calling `Build` (which calls `CalculateSystemContributions`) internally .... as you pointed out in a different PR .. this can be expensive .. !

**Comment by RiccardoRossi** [comment]
i have mixed feelings about this correction. I mean, **we surely have a bug that should be fixed**, however in a sense the application of the constraints is similar to the application of the dirichlet conditions: a different operation than the FEM assembly.

In this sense i believe it is correct to make it independent of the "build", which is pure FEM assembly.

my impression here is that it should be the eigenvalue strategy to call the ApplyConstraints method  (same way as i guess they do for dirichlet conditions) rather than including it in the build...

@adityaghantasala regarding BuildLHS you are surely right that we should have it separate from Build

the version "CompleteOnFreeRows" has never been used (to the best of my knowledge) so i guess we could remove it

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
blocking until we take a decision on what the behaviour should be

**Comment by philbucher** [comment]
I agree with Riccardo that ApplyConstraints should be a separate function 
Then I will add the interface also to the baseclass ok?

**Comment by adityaghantasala** [comment]
The applyconstraints function is currently protected ... it has to be public then .. !

**Comment by RiccardoRossi** [comment]
@adityaghantasala you are right (and i also acknowledge that i was against this :-D)

**Comment by philbucher** [comment]
done, now `ApplyConstraints` is in the Interface of the BuilderAndSolver

**Comment by adityaghantasala** [comment]
Shall we do it in two different PRs ... one onyl with the changes to the B&S .. and the second one with the eigen values stragegy? Then its more convenient to follow the changes later on ?

**Comment by philbucher** [comment]
I would leave it in one PR, then we have everything together and can easily see why the changes were necessary

**Comment by philbucher** [comment]
after discussing in person @adityaghantasala and i decided to make separate PRs, now this PR only contains the changes in the Core

See #5514 for the usage of the new functionalities

ping @RiccardoRossi

**Comment by philbucher** [comment]
image

:D

**Comment by philbucher** [comment]
ping @RiccardoRossi

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 1]
```diff
@@ -871,6 +871,52 @@ class ResidualBasedBlockBuilderAndSolver
         }
     }
 
+    void ApplyConstraints(
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:874]: what is the difference with this? is it only moved within the file?

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:874]: it is moved to `public`

**Review by RiccardoRossi — APPROVED** [review]

**Comment by philbucher** [comment]
thx

**Pull request merged**